hazelcast/include/hazelcast/client/topic/MessageListener.h

```diff
@@ -28,27 +29,50 @@
 namespace hazelcast {
     namespace client {
         namespace topic {
+            class Message;
+            namespace impl {
+                class TopicEventHandlerImpl;
+            }
+
             /**
-             * Invoked when a message is received for the added topic. Note that topic guarantees message ordering.
-             * Therefore there is only one thread invoking onMessage. The user should not keep the thread busy, but preferably
-             * should dispatch it via an Executor. This will increase the performance of the topic.
-             *
-             * @param message the message that is received for the added topic
+             * Listen to messages from an ITopic
+             * \see ITopic::addMessageListener
              */
-            class HAZELCAST_API MessageListener {
+            class HAZELCAST_API MessageListener final {
             public:
-                virtual ~MessageListener() = default;
+                /**
+                 * Set an handler function to be invoked when a message is received for the subscribed topic.
+                 *
+                 * \warning 
+                 * In order to guarantee message ordering, there is only one thread that invokes the given function.
+                 * The user should off-load any time consuming operation to another thread.
+                 *
+                 * \param h a `void` function object that is callable with a single parameter of type `const Message &`
+                 */
+                template<typename Handler,
+                         typename = util::enable_if_rvalue_ref_t<Handler &&>>
+                MessageListener &on_received(Handler &&h) & {
+                    received = std::move(h);
+                    return *this;
+                }
 
                 /**
-                * Invoked when a message is received for the added topic. Note that topic guarantees message ordering.
-                * Therefore there is only one thread invoking onMessage. The user should not keep the thread busy, but preferably
-                * should dispatch it via an Executor. This will increase the performance of the topic.
-                *
-                * @param message the message that is received for the added topic
-                */
-                virtual void onMessage(Message &&message) = 0;
-            };
+                 * \copydoc MessageListener::on_received
+                 */
+                template<typename Handler,
+                         typename = util::enable_if_rvalue_ref_t<Handler &&>>
+                MessageListener &&on_received(Handler &&h) && {
```

Comment:
is this left-over? we already have this method above?

why are you returning rvalue here? I think just lvalue ref would be good enough. returning rvalue reference will also have its own problems https://stackoverflow.com/questions/1116641/is-returning-by-rvalue-reference-more-efficient

Reply:
No, this is not sth accidental, I did the same thing on EntryListener and ItemListener as well. This is because we only accept rvalue parameters in addMessageListener, (or addItemListener and addEntryListener). If you don't return lvalue instead of rvalue ref to *this, then this won't compile:
```c++
topic.addMessageListener(
     MessageListener().
            on_receive(my_func)
)
```
But we and the compiler know that MessageListener() is rvalue here, and when on_receive is called, it will call the second overload with the `) && {`  and that will move the *this back to the caller as an rvalue. If we didnt have this overload than the user would have to use std::move explicitly to explicitly cast the returned lvalue back to rvalue.

Reply:
About the question on StackOverFlow, I don't think we are returning a dangling ref here, we are still returning *this not a copy of *this. The problem there would still exist if the question was asked with lvalues.

Reply:
can you give an example for the first method usage? From your answer i understand that only the second one would have been enough.

Reply:
This wouldn't compile, if we don't overload &:
```
MessageListener l;
l.on_receive(my_func);
topic_addMessageListener(std::move(l));
```
Well, this may look unlikely for the MessageListener, but when there are more events (EntryListener), someone might prefer this.

Reply:
This method of overloading both & and && might look like a code repeat, but it actually isn't. I am calling the & overloaded version from the &&. So, what you implemented in the & version, applies to the && version as well. && version will always be 2 lines: call the & version, and return rvalue ref to *this. Even if we needed to sth more complex in these setter methods (e.g. calculating listener flags in EntryListener), that thing would only be implemented once in the & overload, and the && would automatically use it.